internal/fs/parallel_dirops_test.go

```diff
@@ -0,0 +1,687 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// A collection of tests for a file system where parallel dirops are allowed.
+// Dirops refers to readdir and lookup operations.
+
+package fs_test
+
+import (
+	"os"
+	"path"
+	"sync"
+	"testing"
+
+	"github.com/googlecloudplatform/gcsfuse/v2/internal/config"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/suite"
+)
+
+type ParallelDiropsTest struct {
+	suite.Suite
+	fsTest
+}
+
+type ParallelDiropsWithoutCachesTest struct {
+	ParallelDiropsTest
+}
+
+// createFilesAndDirStructureInBucket creates the following files and directory
+// structure.
+// bucket
+//
+//			file1.txt
+//			file2.txt
+//			explicitDir1
+//					file1.txt
+//					file2.txt
+//	    implicitDir1
+//					file1.txt
+func (t *ParallelDiropsTest) createFilesAndDirStructureInBucket() {
+	assert.Nil(
+		t.T(),
+		t.createObjects(
+			map[string]string{
+				"file1.txt":              "abcdef",
+				"file2.txt":              "xyz",
+				"explicitDir1/":          "",
+				"explicitDir1/file1.txt": "12345",
+				"explicitDir1/file2.txt": "6789101112",
+				"implicitDir1/file1.txt": "-1234556789",
+			}))
+}
+
+func (t *ParallelDiropsTest) SetupSuite() {
+	t.serverCfg.ImplicitDirectories = true
+	t.serverCfg.MountConfig = &config.MountConfig{
+		FileSystemConfig: config.FileSystemConfig{
+			DisableParallelDirops: false,
+		}}
+	t.serverCfg.RenameDirLimit = 10
+	t.fsTest.SetUpTestSuite()
+}
+
+func (t *ParallelDiropsWithoutCachesTest) SetupSuite() {
+	t.serverCfg.ImplicitDirectories = true
+	t.serverCfg.MountConfig = &config.MountConfig{
+		FileSystemConfig: config.FileSystemConfig{
+			DisableParallelDirops: false,
+		}}
+	t.serverCfg.RenameDirLimit = 10
+	t.serverCfg.DirTypeCacheTTL = 0
+	t.serverCfg.InodeAttributeCacheTTL = 0
+	t.fsTest.SetUpTestSuite()
+}
+
+func (t *ParallelDiropsTest) SetupTest() {
+	t.createFilesAndDirStructureInBucket()
+}
+
+func (t *ParallelDiropsTest) TearDownTest() {
+	t.fsTest.TearDown()
+}
+
+func (t *ParallelDiropsTest) TearDownSuite() {
+	t.fsTest.TearDownTestSuite()
+}
+
+func (t *ParallelDiropsTest) TestParallelLookUpsForSameFile() {
+	lookUpFunc := func(wg *sync.WaitGroup, filePath string) (os.FileInfo, error) {
+		defer wg.Done()
+		fileInfo, err := os.Stat(filePath)
+		return fileInfo, err
+	}
+	var stat1, stat2 os.FileInfo
+	var err1, err2 error
+
+	// Parallel lookups of file just under mount.
+	filePath := path.Join(mntDir, "file1.txt")
+	wg := sync.WaitGroup{}
+	wg.Add(2)
+	go func() {
+		stat1, err1 = lookUpFunc(&wg, filePath)
+	}()
+	go func() {
+		stat2, err2 = lookUpFunc(&wg, filePath)
+	}()
+	wg.Wait()
+
+	// Assert both stats passed and give correct information
+	assert.NoError(t.T(), err1)
```

Comment:
Should we create a local function for all these checks and that will be used below?

Reply:
Trying to follow DAMP in unit tests: https://stackoverflow.com/questions/6453235/what-does-damp-not-dry-mean-when-talking-about-unit-tests. 

I know it's a not a good idea to always use DAMP but I think we can use DAMP here without making too much repetitions. LMK if you think otherwise and I will make changes.

Same reason for the second similar comment.

Reply:
Got it. 

Make sense to use DAMP here, since not much repetitions, and readable test.